### PR TITLE
Allow user to turn on zlib, shuffle, and/or fletcher32 filters with parallel I/O for HDF5-1.10.2+

### LINKED ISSUE #1580: Allow parallel I/O writes to use zlib
Allow users to use zlib on parallel writes. This feature is needed urgently by NOAA. No doubt it is desired by many other large data producers.

Large data producers like NOAA are turning to netCDF-4 due to built-in compression. On HPC applications especially, they are writing very large files (10s of GB per file is normal, even with compression).

Since these files are produced on a HPC system, the data start on many processors. Due to the current limitation of the netcdf-c library, they must move all data to one processor, and have that one processor do all compression and writing sequentially. This slows down their very expensive machine!

More importantly, it means that compromises must be made in what is saved. Instead of saving every value that the science team wants, only a subset of values and data are saved. This makes the science harder but makes operations work.

Allowing users to write data compressed with parallel I/O will save them the step of moving all data to one processor, and will also allow the compression to be spread to many processors, instead of just one. This will improve their write time by almost an order of magnitude.

Decrease in I/O time leads to increased time for computation, allowing the science team to increase resolution or use more intensive algorithms, and still meet operational requirements. 

Thus improving the I/O times directly impacts the science.

### PR DESCRIPTION
**Pull request by edwardhartnett**
Fixes #1580

This is the complete PR, with documentation changes, and Cmake and autotools support for determining if the user has access to HDF5 1.10.2.

If the user does not have 1.10.2, they will continue to see NC_EINVAL if they try to set shuffle, zlib, or fletch32 filters. If they have HDF5-1.10.2 or better, then instead of returning an error, these filters will work.

I have change documentation everywhere to reflect and explain the changes.

NOAA is particularly eager for this feature, but many other large data producers will welcome it.

### CODE REVIEW

File: libhdf5/hdf5var.c [hunk 1]
```diff
@@ -661,10 +661,13 @@ nc_def_var_extra(int ncid, int varid, int *shuffle, int *deflate,
         return NC_ENOTVAR;
     assert(var && var->hdr.id == varid);
 
-    /* Can't turn on parallel and deflate/fletcher32/szip/shuffle (for now). */
+    /* Can't turn on parallel and deflate/fletcher32/szip/shuffle
+     * before HDF5 1.10.2. */
+#ifndef HDF5_1_10_2
```
Comment (gsjaardema) [inline libhdf5/hdf5var.c]: Can we use a different more informative symbol.  Maybe `HDF_SUPPORTS_PARALLEL_COMPRESSION`  I think the `HDF5_1_10_2` is confusing especially 5 years from now when we are at 1.14.8

Comment (edwardhartnett) [reply libhdf5/hdf5var.c]: OK, stand by...

**Comment by edwardhartnett** [comment]
@WardF normally I would like you to merge my PRs in the order I write them, but in this case it would be great to skip ahead and merge this one. It will make a great holiday gift to NOAA and other data producers around the world. ;-)

Happy holidays to all at Unidata and throughout the netCDF community! ;-)

**Comment by edhartnett** [comment]
@WardF can we get this merged soon please? The NOAA GFS people will build it and test to ensure this problem is resolved. Then they will know that, starting with the next release of netcdf-c, they will be able to do parallel writes of compressed data...

**Pull request merged**